File: longestUncommonSubsequenceII/longestUncommonSubsequenceII.cpp
// clang++ splitArrayIntoConsecutiveSubsequences2.cpp -std=c++11

#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Comp {
public:
    bool operator()(const string& s1, const string& s2) const {
        if (s1.length() == s2.length())
            return s1 > s2;
        return s1.length() > s2.length();
    }
};
class Solution {
public:
    int findLUSlength(vector<string>& strs) {
        sort(strs.begin(), strs.end(), Comp());
        unordered_set<string> s;
        for (int i = 0; i < strs.size(); i++) {
            if (s.count(strs[i]))
                continue;
            bool isSub = false;
            if (i == strs.size() - 1 || strs[i] != strs[i + 1]) {
                for (string str : s) {
                    int j = 0;
                    for (char c : str) {
                        if (c == strs[i][j])
                            j++;
                        if (j == strs[i].size())
                            break;
                    }
                    if (j == strs[i].size()) {
                        isSub = true;
                        break;
                    }
                }
                if (!isSub)
                    return strs[i].size();
            }
            s.insert(strs[i]);
        } // for
        return -1;
    }
};
int main()
{
  Solution s;
  vector<string> strs = {"aba", "cdc", "eae"};
  cout << s.findLUSlength(strs);
  cout << endl;
  return 0;
}

File: longestUncommonSubsequenceII/longestUncommonSubsequenceII.java
class Solution {
    public int findLUSlength(String[] strs) {
        Arrays.sort(strs, (s1, s2) -> { 
            if (s1.length() == s2.length()) {
                return s1.compareTo(s2);
            } else {
                return s2.length() - s1.length();
            }
        });
        Set<String> s = new HashSet<>();
        for (int i = 0; i < strs.length; i++) {
            if (i + 1 < strs.length && strs[i + 1].equals(strs[i])) {
                s.add(strs[i]);
                continue;
            }
            if (s.contains(strs[i])) {
                continue;
            }
            if (isOneSub(s, strs[i])) {
                s.add(strs[i]);
                continue;
            } else {
                return strs[i].length();
            }
        }
        return -1;
    }
    
    private boolean isOneSub(Set<String> s, String str) {
        for (String str1 : s) {
            if (isSubsequence(str1, str)) {
                return true;
            }
        }
        return false;
    }
    
    private boolean isSubsequence(String s1, String s2) {
        int p1 = 0, p2 = 0;
        while (p1 < s1.length() && p2 < s2.length()) {
            if (s1.charAt(p1) == s2.charAt(p2)) {
                p1++; p2++;
            } else {
                p1++;
            }
        }
        return p2 == s2.length();
    }
}
